### PR TITLE
DAOS-3413 control: server system fanout using new client api

### PR DESCRIPTION
**Pull request by tanabarr**
Update gRPC fanout operations triggered by start/stop/query dmg system
commands that spawn new gRPCs across hosts in the system membership.
Switch to the new control client API for fanout and simplify unit tests
by mocking only specific `UnaryInvoker` methods as opposed to the entire
`MgmtSvcClient` as done previously.

- replace `HarnessClient` implementation by adding with
  `(Start|Stop|...)Ranks` unaryInvoker fanout functions
- simplify `System(Start|Stop|...)` `MgmtSvc` gRPC methods by DRYing
  code and improving rank list filtering
- move instance specific dRPC call handling to `instance_drpc`

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/server/mgmt_client.go [hunk 1]
```diff
@@ -41,37 +41,75 @@ const (
...
+	}
+	// GrpcClient provides a subset of MgmtSvcClient for operations between gRPC servers.
+	GrpcClient interface {
+		GrpcClientCfgHelper
+		Join(context.Context, *mgmtpb.JoinReq) (*mgmtpb.JoinResp, error)
+		PrepShutdown(context.Context, string, mgmtpb.RanksReq) (*mgmtpb.RanksResp, error)
+		Stop(context.Context, string, mgmtpb.RanksReq) (*mgmtpb.RanksResp, error)
+		Start(context.Context, string, mgmtpb.RanksReq) (*mgmtpb.RanksResp, error)
+		Query(context.Context, string, mgmtpb.RanksReq) (*mgmtpb.RanksResp, error)
+	}
```
Comment (mjmac) [inline src/control/server/mgmt_client.go]: From a naming standpoint, "GrpcClient" is a bit vague. How about "SystemClient", or maybe "HarnessClient"? Just something to clarify what kind of client this is.

File: src/control/server/mgmt_client.go [hunk 2]
```diff
@@ -41,37 +41,75 @@ const (
...
+func (gc *grpcClient) delayRetry(ctx context.Context) {
 	select {
 	case <-ctx.Done(): // break early if the parent context is canceled
 	case <-time.After(retryDelay): // otherwise, block until after the delay duration
 	}
 }
 
-func (msc *mgmtSvcClient) withConnection(ctx context.Context, ap string,
+func (gc *grpcClient) withConnection(ctx context.Context, ap string,
 	fn func(context.Context, mgmtpb.MgmtSvcClient) error) error {
```
Comment (mjmac) [inline src/control/server/mgmt_client.go]: If you change the signature of this closure to accept the smaller interface defined above, it should be possible to test all of this code using the mock client that you've added along with the new interface.

**Comment by tanabarr** [comment]
> I took a look at [d0c1f64](https://github.com/daos-stack/daos/commit/d0c1f64186dc040a14c4805a59877bc6716bf627), which is the commit that starts to use the new control API surface. This isn't really what I had in mind. My goal in designing the control API was that the public API methods would be used by all control plane clients, including the servers when they are fanning out from an AP. I wasn't really intending to build out new internal API surface within the server package using primitives from within the control package, which this commit seems to be doing.
> 
> To put it another way, the server package needs to present a gRPC server, but it shouldn't necessarily need to know anything about gRPC clients in order to talk to other servers. If it does, then I would strongly advise that we address any deficiencies in the control API functions. Being our own customer ensures that the API is robust and generally useful.
> 
> I created a draft PR (#2441) to demonstrate the direction that I'm hoping this work will take.
> 
> I think we should use this as an opportunity to step back and revisit some of the design choices made before, rather than just trying to jam this new part into the existing machinery. For example, the logic for handling `client -> AP -> [ranks fanout] -> AP -> client` (in the case of system start/stop) is pretty convoluted and hard to follow. My hope is that we can dramatically simplify this area of the harness code by getting rid of the mgmtClient stuff that was bolted on to the harness instances.

in your draft PR there is significant confusion about how things previously worked, dmg calls control.SystemStart() then you have implemented a call to the same method from within the respective server side handler. that isn't going to work.

**Comment by tanabarr** [comment]
Regarding the comment “For example, the logic for handling client -> AP -> [ranks fanout] -> AP -> client (in the case of system start/stop) is pretty convoluted and hard to follow.”  The flow doesn’t go through the AP after the “ranks fanout”. Just to clarify.

I will get the PR in a functional state and hopefully the intent will be clear.

File: src/control/lib/control/system.go [hunk 3]
```diff
@@ -201,3 +203,75 @@ func ListPools(ctx context.Context, rpcClient UnaryInvoker, req *ListPoolsReq) (
...
+		if srr.HostErrors == nil {
+			srr.HostErrors = make(HostErrorsMap)
+		}
+		return srr.HostErrors.Add(hr.Addr, err)
+	}
+
+	if srr.RankResults == nil {
+		srr.RankResults = make(system.MemberResults, 0)
+	}
+	srr.RankResults = append(srr.RankResults, memberResults...)
```
Comment (mjmac) [inline src/control/lib/control/system.go:238]: For the record, append with a nil slice works fine: https://play.golang.org/p/u3Bnmwrs_sp

This pattern makes sense in the other `addHostResponse()` helpers because we have to make the map before inserting anything.

File: src/control/server/ctl_system.go [hunk 4]
```diff
@@ -231,6 +233,31 @@ func (svc *ControlService) prepShutdown(ctx context.Context, rankList []system.R
...
+	// add rank results for any on reachable harness addrs
+	for errMsg, hostSet := range hostErrors {
+		if !strings.Contains(errMsg, "connection refused") {
+			continue
+		}
+		for _, addr := range strings.Split(hostSet.DerangedString(), ",") {
+			ranks = append(ranks, hostRanks[addr]...)
+			svc.log.Debugf("no response from harness %s", addr)
+		}
+	}
```
Comment (mjmac) [inline src/control/server/ctl_system.go:249]: I don't understand why it matters what the error was. Shouldn't we just be excluding all ranks that are associated with a host error from being in the stopped state?

File: src/control/lib/control/system.go (hunk 3 above)
Comment (tanabarr) [reply src/control/lib/control/system.go]: done

File: src/control/server/ctl_system.go (hunk 4 above)
Comment (tanabarr) [reply src/control/server/ctl_system.go]: the thinking is that there could be another source of error that doesn't indicate that the rank's process is down necessarily, at this point I'm happy to assume a rank is down if any host error is returned.

File: src/control/server/mgmt_client.go [hunk 5]
```diff
@@ -70,17 +70,17 @@ func (msc *mgmtSvcClient) delayRetry(ctx context.Context) {
 }
 
 func withConnection(ctx context.Context, ap string, tc *security.TransportConfig,
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/mgmt_client.go:72]: yes

File: src/control/server/ctl_system_test.go [hunk 6]
```diff
@@ -22,3 +22,106 @@
 //
 
 package server
+
+//const defaultAP = "192.168.1.1:10001"
```
Comment (kjacque) [inline src/control/server/ctl_system_test.go:26]: Going to enable?

File: src/control/server/instance_drpc.go [hunk 7]
```diff
@@ -75,3 +79,92 @@ func (srv *IOServerInstance) CallDrpc(module, method int32, body proto.Message)
...
+	if err != nil {
+		outErr = errors.WithMessagef(err, "rank %s dRPC failed", &rank)
+	} else {
+		resp := &mgmtpb.DaosResp{}
+		if err = proto.Unmarshal(dresp.Body, resp); err != nil {
+			outErr = errors.WithMessagef(err, "rank %s dRPC unmarshal failed",
+				&rank)
+		} else if resp.GetStatus() != 0 {
+			outErr = errors.Errorf("rank %s dRPC returned DER %d",
+				&rank, resp.GetStatus())
```
Comment (kjacque) [inline src/control/server/instance_drpc.go:100]: Are we able to translate the numeric DER value to a string? I thought we were doing that elsewhere in the control plane.

File: src/control/server/instance_drpc.go [hunk 8]
```diff
@@ -75,3 +79,92 @@ func (srv *IOServerInstance) CallDrpc(module, method int32, body proto.Message)
...
+
+	resChan := make(chan *system.MemberResult)
+	go func() {
+		dresp, err := srv.CallDrpc(drpc.ModuleMgmt, drpc.MethodPingRank, nil)
+		resChan <- drespToMemberResult(rank, "ping", dresp, err, system.MemberStateReady)
+	}()
+
+	select {
+	case <-ctx.Done():
+		if ctx.Err() == context.DeadlineExceeded {
```
Comment (kjacque) [inline src/control/server/instance_drpc.go:131]: Does the context show up with a timeout set by default? Or am I just missing where it's set for this command? Seems really important for a ping

File: src/control/server/instance_drpc_test.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+				Body:   rb,
+			}
+
+			gotResult := drespToMemberResult(Rank(dRank), "test", resp, tc.inErr, tc.targetState)
+			if diff := cmp.Diff(tc.expResult, gotResult, common.DefaultCmpOpts()...); diff != "" {
+				t.Fatalf("unexpected response (-want, +got)\n%s\n", diff)
+			}
+		})
+	}
+}
```
Comment (kjacque) [inline src/control/server/instance_drpc_test.go:170]: Is it possible to add unit tests for the ping and prep shutdown calls? Not necessarily in this PR if they're complex.

**Comment by mjmac** [comment]
@tanabarr: I think if you merge latest master this PR should get a much cleaner test run. Would prefer to see relatively clean tests given that this PR touches some key functionality.

File: src/control/server/instance_drpc.go (hunk 7 above)
Comment (tanabarr) [reply src/control/server/instance_drpc.go:100]: yes will look into that as part of follow one reformat PR

File: src/control/server/instance_drpc.go (hunk 8 above)
Comment (tanabarr) [reply src/control/server/instance_drpc.go:131]: this is set in the caller

File: src/control/server/instance_drpc_test.go (hunk 9 above)
Comment (tanabarr) [reply src/control/server/instance_drpc_test.go:170]: test coverage has been added elsewhere for the callers in this PR and will be added for these in the follow-on reformat PR

Comment (tanabarr) [reply src/control/server/instance_drpc_test.go:170]: these methods are covered by the `TestMgmtSvc_{Ping,PrepShutdown}Ranks` unit tests in `server/mgmt_system_test.go`

File: src/control/system/faults.go [hunk 10]
```diff
@@ -23,22 +23,27 @@
...
 		"update system member instead of adding",
 	)
-	FaultMemberMissing = systemFault(
+}
+
+func FaultMemberMissing(rank Rank) *fault.Fault {
+	return systemFault(
 		code.SystemMemberMissing,
-		"system member with given rank doesn't exists",
+		fmt.Sprintf("system member with rank %d doesn't exists", rank),
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/system/faults.go:43]: fix in https://github.com/daos-stack/daos/pull/2539
